test/e2e/storage/non_csi_volumes.go

```diff
@@ -0,0 +1,1170 @@
+	Expect(err).NotTo(HaveOccurred())
+
+	// Parse 'id'' from stdout. Expected format:
+	// |     attachments     |                  []                  |
+	// |  availability_zone  |                 nova                 |
+	// ...
+	// |          id         | 1d6ff08f-5d1c-41a4-ad72-4ef872cae685 |
+	c.volumeID = ""
+	for _, line := range strings.Split(outputString, "\n") {
+		fields := strings.Fields(line)
+		if len(fields) != 5 {
+			continue
+		}
+		if fields[1] != "id" {
+			continue
+		}
+		c.volumeID = fields[3]
+		break
+	}
+	framework.Logf("Volume ID: %s", c.volumeID)
+	Expect(c.volumeID).NotTo(Equal(""))
+}
+
+func (c *cinderDriver) cleanupDriver() {
+	DeleteCinderVolume(c.volumeName)
+}
+
+func DeleteCinderVolume(name string) error {
+	// Try to delete the volume for several seconds - it takes
+	// a while for the plugin to detach it.
+	var output []byte
+	var err error
+	timeout := time.Second * 120
+
+	framework.Logf("Waiting up to %v for removal of cinder volume %s", timeout, name)
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(5 * time.Second) {
+		output, err = exec.Command("cinder", "delete", name).CombinedOutput()
+		if err == nil {
+			framework.Logf("Cinder volume %s deleted", name)
+			return nil
+		} else {
+			framework.Logf("Failed to delete volume %s: %v", name, err)
+		}
+	}
+	framework.Logf("Giving up deleting volume %s: %v\n%s", name, err, string(output[:]))
+	return err
+}
+
+// GCE
+type gceDriver struct {
+	volumeName string
+
+	driverInfo driverInfo
+}
+
+func initGceDriver(f *framework.Framework, config framework.VolumeTestConfig, node v1.Node) testDriver {
+	framework.SkipUnlessProviderIs("gce", "gke")
+
+	ns := f.Namespace
+	fsGroup := int64(1234)
+
+	// PD will be created in framework.TestContext.CloudConfig.Zone zone,
+	// so pods should be also scheduled there.
+	config.NodeSelector = map[string]string{
+		kubeletapis.LabelZoneFailureDomain: framework.TestContext.CloudConfig.Zone,
+	}
+
+	return &gceDriver{
+		driverInfo: driverInfo{
+			name:    "gce",
+			fsGroup: &fsGroup,
+			podSec: v1.PodSecurityContext{
+				FSGroup: &fsGroup,
+			},
+			privilegedSecurityContext: true,
+			maxFileSize:               fileSizeMedium,
+			isBlockSupported:          false,
+			expectedContent:           "Hello from GCE from namespace " + ns.Name,
+			needInjectHTML:            true,
+			testFile:                  "gce_io_test",
+			f:                         f,
+			config:                    config,
+			node:                      node,
+		},
+	}
+}
+
+func (g *gceDriver) getDriverInfo() driverInfo {
+	return g.driverInfo
+}
+
+func (g *gceDriver) getVolumeSource(readOnly bool) *v1.VolumeSource {
+	return &v1.VolumeSource{
+		GCEPersistentDisk: &v1.GCEPersistentDiskVolumeSource{
+			PDName:   g.volumeName,
+			FSType:   "ext3",
+			ReadOnly: readOnly,
+		},
+	}
+}
+
+func (g *gceDriver) getPersistentVolumeSource(readOnly bool) *v1.PersistentVolumeSource {
+	return nil
```

Comment:
we primarily want to test gce pd with PV.  You can copy the GCE PD PV setup from the subpath test.